File: src/chip_8/delay_timer.rs
pub struct DelayTimer {
    value: u8
}

impl DelayTimer {
    pub fn new() -> DelayTimer {
        DelayTimer {
            value: 0
        }
    }
}

impl crate::traits::Timer for DelayTimer {
    fn value(&self) -> u8 {
        self.value
    }

    fn set_value(&mut self, value: u8) {
        self.value = value;
    }

    fn update(&mut self) -> Result<(), String> {
        if self.value > 0 {
            self.value -= 1;
        }
        Ok(())
    }
}

File: wasm/www/components/DelayTimer.js
import { renderNumberValue } from './Utility'

class DelayTimer extends HTMLElement {
    constructor() {
      super();

      this.value = 0;
    }
  
    connectedCallback() {
      this.render();
    }

    update(new_value) {
        if (this.value == new_value) {
            return;
        }
        
        if (typeof new_value != 'number') {
            console.error("Unexpected delay timer value type: ", typeof new_value);
            return;
        }
        
        this.value = new_value;

        this.render();
    }
  
    render() {
      this.innerHTML = `
        <div class="contents">
            ${renderNumberValue("DELAY", this.value, 2)}
        </div>
        `;
    }
  }
  
  customElements.define("delay-timer", DelayTimer);